ggml/src/ggml-vulkan/ggml-vulkan.cpp

```diff
@@ -2743,44 +2996,12 @@ static uint32_t ggml_vk_guess_split_k(int m, int n, int k) {
     GGML_UNUSED(m); GGML_UNUSED(n); GGML_UNUSED(k);
 }
 
-static vk_pipeline ggml_vk_guess_matmul_pipeline_amd(ggml_backend_vk_context * ctx, vk_matmul_pipeline& mmp, int m, int n, bool aligned) {
-    if (m <= 32 || n <= 32) {
-        return aligned ? mmp->a_s : mmp->s;
-    }
-    return aligned ? mmp->a_m : mmp->m;
-
-    GGML_UNUSED(ctx);
-}
-
-static vk_pipeline ggml_vk_guess_matmul_pipeline_apple(ggml_backend_vk_context * ctx, vk_matmul_pipeline& mmp, bool aligned) {
-    return aligned ? mmp->a_m : mmp->m;
-
-    GGML_UNUSED(ctx);
-}
-
-static vk_pipeline ggml_vk_guess_matmul_pipeline_intel(ggml_backend_vk_context * ctx, vk_matmul_pipeline& mmp, bool aligned) {
-    return aligned ? mmp->a_s : mmp->s;
-
-    GGML_UNUSED(ctx);
-}
-
 static vk_pipeline ggml_vk_guess_matmul_pipeline(ggml_backend_vk_context * ctx, vk_matmul_pipeline& mmp, int m, int n, bool aligned) {
     VK_LOG_DEBUG("ggml_vk_guess_matmul_pipeline(" << m << ", " << n << ", " << aligned << ")");
-    switch (ctx->device->vendor_id) {
-    case VK_VENDOR_ID_AMD:
-        return ggml_vk_guess_matmul_pipeline_amd(ctx, mmp, m, n, aligned);
-    case VK_VENDOR_ID_APPLE:
-        return ggml_vk_guess_matmul_pipeline_apple(ctx, mmp, aligned);
-    case VK_VENDOR_ID_INTEL:
-        return ggml_vk_guess_matmul_pipeline_intel(ctx, mmp, aligned);
-    default:
-        break;
-    }
-
-    if (m <= 32 || n <= 32) {
+    if ((ctx->device->mul_mat_s && (m <= 32 || n <= 32)) || (!ctx->device->mul_mat_m && !ctx->device->mul_mat_l)) {
         return aligned ? mmp->a_s : mmp->s;
     }
-    if (m <= 64 || n <= 64) {
+    if ((ctx->device->mul_mat_m && (m <= 64 || n <= 64 || ctx->device->coopmat_support)) || !ctx->device->mul_mat_l) {
```

Comment:
I don't think coopmat_support should force medium, at least not on NVIDIA.

Reply:
When it first started working I ran the three tile sizes next to each other and the large tile was very slow. I'll test it again.

Reply:
It's not as slow as I remember, but performance drops significantly on RTX 3090 if I allow the large tile.

Reply:
I narrowed it down a little further: Performance drops significantly with the large tiles when working with quants. With float16 or float32 the large tiles work well with coopmat.

Reply:
You... are right. So basically just unroll every loop you can get your hands on?

I guess it's dynamic access to shared memory this time. Why did that affect the large tile so significantly?

Reply:
I think it's still the sums array, but maybe it was getting unrolled automatically for the medium size but not for the large. The compilers aren't as aggressive about this as they probably should be, so anytime you have a dynamically indexed array where unrolling would make the indices constant, it's usually worth doing (and that's why `#pragma unroll` shows up literally hundreds of times in the cuda kernels).